processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java

```diff
@@ -714,14 +749,20 @@ public int getOffset()
     @Override
     public Offset clone()
     {
-      return new TimestampCheckingOffset(baseOffset.clone(), timestamps, threshold, allWithinThreshold);
+      return new TimestampCheckingOffset(baseOffset.clone(), timestamps, timeStart, timeEnd, allWithinThreshold);
     }
 
     @Override
     public boolean withinBounds()
     {
-      return baseOffset.withinBounds() && (allWithinThreshold
-                                           || timestamps.getLongSingleValueRow(baseOffset.getOffset()) < threshold);
+      if (!baseOffset.withinBounds()) {
+        return false;
+      }
+      if (allWithinThreshold) {
+        return true;
+      }
+      long current = timestamps.getLongSingleValueRow(baseOffset.getOffset());
+      return current >= timeStart && current < timeEnd;
```

Comment:
given that we technically only need to check one of those two conditions based on whether the query is descending or not, is it faster to do a check based on the descending flag, to always check both, or is there maybe a benefit to do the branching outside of the loop, i.e have something like a `DescendingTimestampCheckingOffset?`